public/friendlycode/js/fc/filesystem-sync.js

```diff
@@ -103,22 +114,124 @@ define(["jquery"], function($) {
 
     function configHandler(handler, url) {
       return function() {
-        if(fsync.beforeEach) {
-          fsync.beforeEach();
-        }
+        trigger(fsync._callbacks.beforeEach, arguments);
 
         Array.prototype.unshift.call(arguments, url, csrfToken, fs);
         handler.apply(fsync, arguments);
       };
     }
     Bramble.once("ready", function(bramble) {
-      bramble.on("fileChange", configHandler(pushFileChange, persistanceUrls.createOrUpdate));
-      bramble.on("fileDelete", configHandler(pushFileDelete, persistanceUrls.del));
-      bramble.on("fileRename", configHandler(pushFileRename, persistanceUrls.rename));
+      fsync.bramble = bramble;
+      fsync.handlers = {
+        change: configHandler(pushFileChange, persistanceUrls.createOrUpdate),
+        del: configHandler(pushFileDelete, persistanceUrls.del),
+        rename: configHandler(pushFileRename, persistanceUrls.rename)
+      };
+
+      bramble.on("fileChange", fsync.handlers.change);
+      bramble.on("fileDelete", fsync.handlers.del);
+      bramble.on("fileRename", fsync.handlers.rename);
     });
 
     return fsync;
   };
 
+  FSync.prototype.saveAndSyncAll = function(callback) {
+    var fsync = this;
+    var bramble = fsync.bramble;
+    var afterEach = fsync._callbacks.afterEach;
+    var filesSaved = [];
+
+    function resetAfterEach(fn) {
+      var index = afterEach.indexOf(fn);
+      if(index !== -1) {
+        afterEach.splice(index, 1);
+      }
+    }
+
+    function maybeFinish(err, path) {
+      function finish() {
+        resetAfterEach(maybeFinish);
+        bramble.on("fileChange", fsync.handlers.change);
+        callback.apply(null, arguments);
+      }
+
+      if(err) {
+        finish(err, path);
+        return;
+      }
+
+      if(path) {
+        filesSaved.splice(filesSaved.indexOf(path), 1);
+      }
+
+      if(filesSaved.length === 0) {
+        finish();
+      }
+    }
+
+    function syncAll() {
+      if(filesSaved.length === 0) {
+        maybeFinish();
+        return;
+      }
+
+      fsync.afterEach(maybeFinish);
+      filesSaved.forEach(function(path) {
+        fsync.handlers.change(path);
+      });
+    }
+
+    function cacheFilesSaved(path) {
```

Comment:
I wonder if this name should indicate that you're caching a file *path* and not the file itself, so there's no confusion.